Race condition in paths.get_ipython_dirs()

Mirroring the comments in https://github.com/ipython/ipython/commit/7a20999bb2097a781b81e3098ba3e379cdd6b14e

I believe this commit is causing downstream breakage in Treon (a simultaneous jupyter notebook test runner). In short, it's a race condition where if the directory doesn't exist, it may have been created in between the time of test and time of attempted creation. This happens when multiple tests are run at the same time, and they all hit this codepath at the same time.

Stack trace:

```
Traceback (most recent call last):
  File "/usr/share/miniconda/lib/python3.7/site-packages/traitlets/traitlets.py", line 537, in get
    value = obj._trait_values[self.name]
KeyError: 'kernel_dirs'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/share/miniconda/bin/treon", line 8, in <module>
    sys.exit(main())
  File "/usr/share/miniconda/lib/python3.7/site-packages/treon/treon.py", line 52, in main
    trigger_tasks(tasks, thread_count)
  File "/usr/share/miniconda/lib/python3.7/site-packages/treon/treon.py", line 71, in trigger_tasks
    pool.map(Task.run_tests, tasks)
  File "/usr/share/miniconda/lib/python3.7/multiprocessing/pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/share/miniconda/lib/python3.7/multiprocessing/pool.py", line 657, in get
    raise self._value
  File "/usr/share/miniconda/lib/python3.7/multiprocessing/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/usr/share/miniconda/lib/python3.7/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "/usr/share/miniconda/lib/python3.7/site-packages/treon/task.py", line 25, in run_tests
    self.is_successful, console_output = execute_notebook(self.file_path)
  File "/usr/share/miniconda/lib/python3.7/site-packages/treon/test_execution.py", line 13, in execute_notebook
    processor.preprocess(notebook, metadata(path))
  File "/usr/share/miniconda/lib/python3.7/site-packages/nbconvert/preprocessors/execute.py", line 80, in preprocess
    with self.setup_kernel():
  File "/usr/share/miniconda/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/usr/share/miniconda/lib/python3.7/site-packages/nbclient/client.py", line 456, in setup_kernel
    self.start_new_kernel(**kwargs)
  File "/usr/share/miniconda/lib/python3.7/site-packages/nbclient/util.py", line 78, in wrapped
    return just_run(coro(*args, **kwargs))
  File "/usr/share/miniconda/lib/python3.7/site-packages/nbclient/util.py", line 57, in just_run
    return loop.run_until_complete(coro)
  File "/usr/share/miniconda/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "/usr/share/miniconda/lib/python3.7/site-packages/nbclient/client.py", line 412, in async_start_new_kernel
    await ensure_async(self.km.start_kernel(extra_arguments=self.extra_arguments, **kwargs))
  File "/usr/share/miniconda/lib/python3.7/site-packages/nbclient/util.py", line 89, in ensure_async
    result = await obj
  File "/usr/share/miniconda/lib/python3.7/site-packages/jupyter_client/manager.py", line 332, in _async_start_kernel
    kernel_cmd, kw = await ensure_async(self.pre_start_kernel(**kw))
  File "/usr/share/miniconda/lib/python3.7/site-packages/jupyter_client/utils.py", line 34, in ensure_async
    return await obj
  File "/usr/share/miniconda/lib/python3.7/site-packages/jupyter_client/manager.py", line 296, in _async_pre_start_kernel
    self.kernel_spec,
  File "/usr/share/miniconda/lib/python3.7/site-packages/jupyter_client/manager.py", line 131, in kernel_spec
    self._kernel_spec = self.kernel_spec_manager.get_kernel_spec(self.kernel_name)
  File "/usr/share/miniconda/lib/python3.7/site-packages/jupyter_client/kernelspec.py", line 289, in get_kernel_spec
    resource_dir = self._find_spec_directory(kernel_name.lower())
  File "/usr/share/miniconda/lib/python3.7/site-packages/jupyter_client/kernelspec.py", line 264, in _find_spec_directory
    for kernel_dir in [kd for kd in self.kernel_dirs if os.path.isdir(kd)]:
  File "/usr/share/miniconda/lib/python3.7/site-packages/traitlets/traitlets.py", line 577, in __get__
    return self.get(obj, cls)
  File "/usr/share/miniconda/lib/python3.7/site-packages/traitlets/traitlets.py", line 540, in get
    default = obj.trait_defaults(self.name)
  File "/usr/share/miniconda/lib/python3.7/site-packages/traitlets/traitlets.py", line 1580, in trait_defaults
    return self._get_trait_default_generator(names[0])(self)
  File "/usr/share/miniconda/lib/python3.7/site-packages/jupyter_client/kernelspec.py", line 208, in _kernel_dirs_default
    dirs.append(os.path.join(get_ipython_dir(), "kernels"))
  File "/usr/share/miniconda/lib/python3.7/site-packages/IPython/paths.py", line 70, in get_ipython_dir
    os.makedirs(ipdir)
  File "/usr/share/miniconda/lib/python3.7/os.py", line 223, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/home/runner/.ipython'
Error: Process completed with exit code 1.
```

--- a/IPython/paths.py
+++ b/IPython/paths.py
@@ -72,7 +72,7 @@ def get_ipython_dir() -> str:
                     " using a temp directory.".format(parent))
             ipdir = tempfile.mkdtemp()
         else:
-            os.makedirs(ipdir)
+            ensure_dir_exists(ipdir)
     assert isinstance(ipdir, str), "all path manipulation should be str(unicode), but are not."
     return ipdir
 

